File: pygame/pipe_spawner.py
#!/usr/bin/env python3
import random
import pygame

from settings import *

from pipe import Pipe

class PipeSpawner:
    # =========================================================
    def __init__(self):
        self.pipe_img_bottom =  pygame.image.load(RESOURSE_PIPE_GREEN_PATH)
        self.pipe_img_top = pygame.transform.flip(self.pipe_img_bottom , False, True)
        self.pipes = pygame.sprite.Group()
        self.is_started = False
        self.pipe_timer = 0
        self.score_listener = None

    # =========================================================
    def update(self, window):
        # Pipe spawner
        if self.is_started:
            if self.pipe_timer <= 0:
                x_top, x_bottom = PIPE_SPAWNER_POS_X, PIPE_SPAWNER_POS_X
                y_top = random.randint(PIPE_SPAWNER_POSY_RAND_RANGE_MIN, PIPE_SPAWNER_POSY_RAND_RANGE_MAX)
                y_bottom = y_top + PIPE_SPAWNER_POSY_GAP + RESOURSE_PIPE_HEIGHT

                pipe_top: Pipe = Pipe(x_top, y_top, self.pipe_img_top, True)
                pipe_bottom: Pipe = Pipe(x_bottom, y_bottom, self.pipe_img_bottom, False)
                pipe_bottom.score_listener = self.score_listener 

                self.pipes.add(pipe_top)
                self.pipes.add(pipe_bottom)

                self.pipe_timer = PIPE_SPAWNER_TIMEOUT
            self.pipe_timer -= 1

        # Update pipe positions
        if self.is_started:
            self.pipes.update()

    # =========================================================
    def draw(self, window):
        self.pipes.draw(window)

    # =========================================================
    def on_start(self):
        self.is_started = True

    # =========================================================
    def on_game_over(self):
        self.is_started = False

File: java.desktop/src/FlappyBird/PipeSpawner.java
package FlappyBird;

import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.image.BufferedImage;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Random;

public class PipeSpawner {
	private final BufferedImage pipeImgBottom;
	private final BufferedImage pipeImgTop;

	private final List<Pipe> pipes = new ArrayList<>();

	private final Random random = new Random();

	private boolean started = false;

	private int pipeTimer = 0;

	private IScoreListener scoreListener;

	public PipeSpawner() {
		pipeImgBottom = ResourceLoader.loadImage(Settings.RESOURCE_PIPE_GREEN_PATH);
		pipeImgTop = ResourceLoader.flipVerticalImage(pipeImgBottom);
	}

	public void update() {
		if (started) {
			if (pipeTimer <= 0) {
				spawnPipePair();

				pipeTimer = Settings.PIPE_SPAWNER_TIMEOUT;
			}

			pipeTimer--;

			Iterator<Pipe> iterator = pipes.iterator();

			while (iterator.hasNext()) {
				Pipe pipe = iterator.next();

				pipe.update();

				if (pipe.outOfWindow()) {
					iterator.remove();
				}
			}
		}
	}

	private void spawnPipePair() {
		int xTop = Settings.PIPE_SPAWNER_POS_X;

		int xBottom = Settings.PIPE_SPAWNER_POS_X;

		int yTop = random
				.nextInt(Settings.PIPE_SPAWNER_POSY_RAND_RANGE_MAX - Settings.PIPE_SPAWNER_POSY_RAND_RANGE_MIN + 1)
				+ Settings.PIPE_SPAWNER_POSY_RAND_RANGE_MIN;

		int yBottom = yTop + Settings.PIPE_SPAWNER_POSY_GAP + Settings.RESOURCE_PIPE_HEIGHT;

		Pipe topPipe = new Pipe(xTop, yTop, pipeImgTop, true);

		Pipe bottomPipe = new Pipe(xBottom, yBottom, pipeImgBottom, false);

		bottomPipe.setScoreListener(scoreListener);

		pipes.add(topPipe);
		pipes.add(bottomPipe);
	}

	public void draw(Graphics g) {
		for (Pipe pipe : pipes) {
			pipe.draw(g);
		}
	}

	public void onStart() {
		started = true;
	}

	public void onGameOver() {
		started = false;
	}
	
	public void onRestart() {
		started = false;
		pipeTimer = 0;
		pipes.clear();
	}

	public void setScoreListener(IScoreListener listener) {
		this.scoreListener = listener;
	}

	public List<Pipe> getPipes() {
		return pipes;
	}

}